src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs

```diff
@@ -247,22 +247,34 @@ public bool TryGetTelemetryId(out Guid telemetryId)
 
                 if (selectionOpt?.Length > 0)
                 {
-                    // There was a selection.  Treat refactorings as more important than 
-                    // fixes.  Note: we still will sort after this.  So any high pri fixes
-                    // will come to the front.  Any low-pri refactorings will go to the end.
+                    // There was a selection.  Treat refactorings as more important than fixes.
+                    // Note: we still will sort after this.  So any high pri fixes will come to the
+                    // front.  Any low-pri refactorings will go to the end.
                     return refactorings.Concat(fixes);
                 }
                 else
                 {
-                    // No selection.  Treat all refactorings as low priority, and place
-                    // after fixes.  Even a low pri fixes will be above what was *originally*
+                    // No selection.  Treat all medium and low pri refactorings as low priority, and
+                    // place after fixes.  Even a low pri fixes will be above what was *originally*
                     // a medium pri refactoring.
-                    refactorings = refactorings.SelectAsArray(r => new SuggestedActionSet(
-                        r.CategoryName, r.Actions, r.Title, SuggestedActionSetPriority.Low, r.ApplicableToSpan));
-                    return fixes.Concat(refactorings);
+                    //
+                    // Note: we do not do this for *high* pri refactorings (like 'rename').  These
+                    // are still very important and need to stay at the top (though still after high
+                    // pri fixes).
+                    var highPriRefactorings = refactorings.WhereAsArray(s => s.Priority == SuggestedActionSetPriority.High);
+                    var nonHighPriRefactorings = refactorings.WhereAsArray(s => s.Priority != SuggestedActionSetPriority.High)
+                                                             .SelectAsArray(s => WithPriority(s, SuggestedActionSetPriority.Low));
+
+                    var highPriFixes = fixes.WhereAsArray(s => s.Priority == SuggestedActionSetPriority.High);
+                    var nonHighPriFixes = fixes.WhereAsArray(s => s.Priority != SuggestedActionSetPriority.High);
+
+                    return highPriFixes.Concat(highPriRefactorings).Concat(nonHighPriFixes).Concat(nonHighPriRefactorings);
```

Comment:
this is the main thrust of hte ordering change.  Here's the *new* logic.

1. fixes and refactorings can specify a priority they want.
2. 'high' priority is *always* respected, for both refactorings and fixes. 
3. high-pri fixes beat high-pri refactorings.  Note: this means `add using` continues to beat `rename` (just like before).
4. medium/low fixes always beat medum/low refactorings.
5. 'rename' is now a high-pri refactoring.

This allows all ordering to effectively stay the same as it was before.  It just prevents 'rename' from being pushed to the bottom since it's considered a very high pri refactoring.  And, like before, it still comes after add-using.

Reply:
This looks good. Just to add, the `CodeAction.Priority` is an internal property, so all third party fixes and refactorings will always be in the default medium priority bucket, and we can choose to promote any of our fixes and refactorings to high or low priority appropriately.
https://github.com/dotnet/roslyn/blob/2f299ee5e5e47b2689b31a25f7788cdcc1d5e646/src/Workspaces/Core/Portable/CodeActions/CodeAction.cs#L56

Reply:
Do we have an test *somewhere* that tests the point you made in 3 that add using still beats rename if you're changing a type to something that's not in scope? (I actually had to go and think whether this even made sense...)